File: python/agentevals/trajectory/tool_permission.py
from typing import Optional, Union

from agentevals.trajectory.utils import (
    _normalize_to_openai_messages_list,
    _extract_tool_calls,
)
from agentevals.types import (
    ChatCompletionMessage,
    SimpleEvaluator,
    SimpleAsyncEvaluator,
)
from agentevals.utils import _run_evaluator, _arun_evaluator

from langchain_core.messages import BaseMessage


def _score_tool_permissions(
    outputs: list[ChatCompletionMessage],
    allowed_tools: Optional[set[str]],
    denied_tools: set[str],
) -> float:
    tool_calls = _extract_tool_calls(outputs)
    if not tool_calls:
        # No tools were called, so no permission boundary could be violated.
        return 1.0
    unauthorized = 0
    for call in tool_calls:
        name = call.get("name")
        if name in denied_tools:
            unauthorized += 1
        elif allowed_tools is not None and name not in allowed_tools:
            unauthorized += 1
    total = len(tool_calls)
    return (total - unauthorized) / total


def _make_scorer(allowed_tools: Optional[set[str]], denied_tools: set[str]):
    def _scorer(*, outputs: list[ChatCompletionMessage], **kwargs) -> float:
        return _score_tool_permissions(outputs, allowed_tools, denied_tools)

    return _scorer


def create_trajectory_tool_permission_evaluator(
    *,
    allowed_tools: Optional[list[str]] = None,
    denied_tools: Optional[list[str]] = None,
) -> SimpleEvaluator:
    """Creates an evaluator that checks whether an agent only called tools it was
    authorized to, based on a permission policy (independent of any reference trajectory).

    Unlike ``create_trajectory_match_evaluator``, which compares called tools against a
    reference trajectory, this evaluator enforces least privilege: it flags any tool call
    outside the granted policy, regardless of whether the task was completed.

    Args:
        allowed_tools (Optional[list[str]]): Allowlist of permitted tool names. If provided,
            any called tool not in this list counts as unauthorized (least privilege).
        denied_tools (Optional[list[str]]): Denylist of forbidden tool names. Any called tool
            in this list counts as unauthorized. A denial always takes precedence over an allow.

    Returns:
        SimpleEvaluator: A function that accepts `outputs` (the agent trajectory) and returns
        an EvaluatorResult whose score is the fraction of tool calls that were authorized
        (1.0 when no tools were called).

    Example:
    ```python
    evaluator = create_trajectory_tool_permission_evaluator(
        allowed_tools=["search_kb", "reply_to_customer"],
        denied_tools=["issue_refund"],
    )
    result = evaluator(outputs=...)
    ```
    """
    if allowed_tools is None and denied_tools is None:
        raise ValueError(
            "create_trajectory_tool_permission_evaluator requires at least one of "
            "`allowed_tools` (an allowlist) or `denied_tools` (a denylist)."
        )
    allowed = set(allowed_tools) if allowed_tools is not None else None
    denied = set(denied_tools or [])
    scorer = _make_scorer(allowed, denied)

    def _wrapped_evaluator(
        *,
        outputs: Union[list[ChatCompletionMessage], list[BaseMessage], dict],
        **kwargs,
    ):
        outputs = _normalize_to_openai_messages_list(outputs)
        return _run_evaluator(
            run_name="trajectory_tool_permission",
            scorer=scorer,
            feedback_key="trajectory_tool_permission",
            outputs=outputs,
            **kwargs,
        )

    return _wrapped_evaluator


def create_async_trajectory_tool_permission_evaluator(
    *,
    allowed_tools: Optional[list[str]] = None,
    denied_tools: Optional[list[str]] = None,
) -> SimpleAsyncEvaluator:
    """Async version of ``create_trajectory_tool_permission_evaluator``."""
    if allowed_tools is None and denied_tools is None:
        raise ValueError(
            "create_async_trajectory_tool_permission_evaluator requires at least one of "
            "`allowed_tools` (an allowlist) or `denied_tools` (a denylist)."
        )
    allowed = set(allowed_tools) if allowed_tools is not None else None
    denied = set(denied_tools or [])
    scorer = _make_scorer(allowed, denied)

    async def _wrapped_evaluator(
        *,
        outputs: Union[list[ChatCompletionMessage], list[BaseMessage], dict],
        **kwargs,
    ):
        outputs = _normalize_to_openai_messages_list(outputs)
        return await _arun_evaluator(
            run_name="trajectory_tool_permission",
            scorer=scorer,
            feedback_key="trajectory_tool_permission",
            outputs=outputs,
            **kwargs,
        )

    return _wrapped_evaluator

File: js/src/trajectory/toolPermission.ts
import { BaseMessage } from "@langchain/core/messages";
import {
  ChatCompletionMessage,
  FlexibleChatCompletionMessage,
} from "../types.js";
import { _normalizeToOpenAIMessagesList, _runEvaluator } from "../utils.js";
import { _extractToolCalls } from "./utils.js";

function _scoreToolPermissions(
  outputs: ChatCompletionMessage[],
  allowedTools: Set<string> | null,
  deniedTools: Set<string>
): number {
  const toolCalls = _extractToolCalls(outputs);
  if (toolCalls.length === 0) {
    // No tools were called, so no permission boundary could be violated.
    return 1;
  }
  let unauthorized = 0;
  for (const call of toolCalls) {
    const name = call.name;
    if (deniedTools.has(name)) {
      unauthorized += 1;
    } else if (allowedTools !== null && !allowedTools.has(name)) {
      unauthorized += 1;
    }
  }
  return (toolCalls.length - unauthorized) / toolCalls.length;
}

/**
 * Creates an evaluator that checks whether an agent only called tools it was authorized to,
 * based on a permission policy (independent of any reference trajectory).
 *
 * Unlike `createTrajectoryMatchEvaluator`, which compares called tools against a reference
 * trajectory, this evaluator enforces least privilege: it flags any tool call outside the
 * granted policy, regardless of whether the task was completed.
 *
 * @param options - The configuration options
 * @param options.allowedTools - Allowlist of permitted tool names. If provided, any called
 *   tool not in this list counts as unauthorized (least privilege).
 * @param options.deniedTools - Denylist of forbidden tool names. Any called tool in this list
 *   counts as unauthorized. A denial always takes precedence over an allow.
 *
 * @returns An async function that accepts `outputs` (the agent trajectory) and returns an
 *   EvaluatorResult whose score is the fraction of tool calls that were authorized
 *   (1 when no tools were called).
 *
 * @example
 * ```typescript
 * const evaluator = createTrajectoryToolPermissionEvaluator({
 *   allowedTools: ["search_kb", "reply_to_customer"],
 *   deniedTools: ["issue_refund"],
 * });
 * const result = await evaluator({ outputs: [...] });
 * ```
 */
export function createTrajectoryToolPermissionEvaluator({
  allowedTools,
  deniedTools,
}: {
  allowedTools?: string[];
  deniedTools?: string[];
}) {
  if (allowedTools == null && deniedTools == null) {
    throw new Error(
      "createTrajectoryToolPermissionEvaluator requires at least one of " +
        "`allowedTools` (an allowlist) or `deniedTools` (a denylist)."
    );
  }
  const allowed = allowedTools != null ? new Set(allowedTools) : null;
  const denied = new Set(deniedTools ?? []);

  const scorer = (params: {
    outputs: ChatCompletionMessage[];
    [key: string]: unknown;
  }): number => _scoreToolPermissions(params.outputs, allowed, denied);

  return async function _wrappedEvaluator({
    outputs,
    ...extra
  }: {
    outputs:
      | ChatCompletionMessage[]
      | FlexibleChatCompletionMessage[]
      | BaseMessage[]
      | {
          messages: (
            | BaseMessage
            | ChatCompletionMessage
            | FlexibleChatCompletionMessage
          )[];
        };
    [key: string]: unknown;
  }) {
    const normalizedOutputs = _normalizeToOpenAIMessagesList(outputs);
    return _runEvaluator(
      "trajectory_tool_permission",
      scorer,
      "trajectory_tool_permission",
      {
        outputs: normalizedOutputs,
        ...extra,
      }
    );
  };
}
